### DISCUSSION TITLE
Please make OptimizerOptions more accessible

### DISCUSSION THREAD

**Discussion by NathanCQC** [NVIDIA/cuQuantum#61] [Q&A] [answered]
The Optimizer options are very very difficult to use. Particularly if you want to control things like slicing across nodes.

Having each thing as its own dictionary or a specific dataclass is very confusing. (I like that they are data classes). But how they are created by

`SlicerOptions = _create_options('SlicerOptions', _opt_conf_enum, _get_dtype, 'slicer', 'SLICER_(?P<option_name>.*)')
`
 It is incredibly un user friendly and hard to follow.

Also this file is very difficult to decipher.

https://github.com/NVIDIA/cuQuantum/blob/2cc68ede865d1f2028f28c9ad411f85748e20b58/python/cuquantum/cutensornet/configuration.py#L101

Also the fact that you can set your own hidden ENUMs with

```
    network.optimizer_config_ptr = cutn.create_contraction_optimizer_config(network.handle)
    enum = cutn.ContractionOptimizerConfigAttribute.SIMPLIFICATION_DISABLE_DR
    value = 1
    name = 'SIMPLIFICATION_DISABLE_DR'

    network._set_opt_config_option(name,enum, value)
```

Is very difficult to use.

There is very little documentation, but it is very important.

Searching OptimizerOptions returns little which is helpful.

https://github.com/search?q=repo%3ANVIDIA%2FcuQuantum%20OptimizerOptions&type=code

The two examples there are very minimal.

**Comment by leofang** [comment, marked as answer]
Hi @NathanCQC thanks for sharing feedbacks! I would like to better understand the challenges that you're facing, answer question you might have, and explore room for potential improvements.

> The Optimizer options are very very difficult to use. Particularly if you want to control things like slicing across nodes.

Could you perhaps elaborate on how it is difficult to use, and, in particular, why it's hard for multi-nodes? The same option should be automatically propagated to all nodes by cuTensorNet if you use the [auto parallelism](https://github.com/NVIDIA/cuQuantum/blob/2cc68ede865d1f2028f28c9ad411f85748e20b58/python/samples/cutensornet/coarse/example22_mpi_auto.py). If [parallelizing manually](https://github.com/NVIDIA/cuQuantum/blob/2cc68ede865d1f2028f28c9ad411f85748e20b58/python/samples/cutensornet/fine/example3_mpi_buffered.py), yes you'd need to ensure the same optimizer option is used by all worker processes for consistency.   

> Having each thing as its own dictionary or a specific dataclass is very confusing. (I like that they are data classes).

I am not sure I follow this statement. Could you elaborate?

Python dataclasses are often interchangeable with dicts, for example there's a built-in [`asdict`](https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict) method. You can also unpack a dict when creating an dataclass instance (ex: `my_dataclass(**dict)`). So what we offer is just to make it a flexible, convenient two-way street for users: We can take either a dict or a dataclass, it's up to the user's choice.

> But how they are created by
> ...
> It is incredibly un user friendly and hard to follow.
> Also this file is very difficult to decipher.

I understand that the readability of the quoted code is not the best, but we need to autogenerate a lot of options at run time to support a wide variety of needs while avoiding boilerplate code that's hard to maintain. In any case, the autogen functionality is considered internal and users should not be bothered with it. 

> Also the fact that you can set your own hidden ENUMs with
> ...
> Is very difficult to use.

The quoted code is, again, internal to our APIs (notice `_set_opt_config_option()` is prefixed by an underscore, following standard Python practice). May I ask what is the reason for accessing internal methods?

> Searching OptimizerOptions returns little which is helpful.

Examples [8](https://github.com/NVIDIA/cuQuantum/blob/2cc68ede865d1f2028f28c9ad411f85748e20b58/python/samples/cutensornet/coarse/example8.py#L12), [12](https://github.com/NVIDIA/cuQuantum/blob/2cc68ede865d1f2028f28c9ad411f85748e20b58/python/samples/cutensornet/coarse/example12.py#L10), [13](https://github.com/NVIDIA/cuQuantum/blob/2cc68ede865d1f2028f28c9ad411f85748e20b58/python/samples/cutensornet/coarse/example13.py#L10) are relevant for using `OptimizerOptions` The options are also [documented here](https://docs.nvidia.com/cuda/cuquantum/python/api/cuquantum.html#objects), with cross referencing set up. 

In case it was unclear, the contraction functionality (ex: `cuquantum.contract()`, `cuquantum.Network`, etc) is built on top of cuTensorNet, so the cuTensorNet documentation (ex: the [Overview](https://docs.nvidia.com/cuda/cuquantum/cutensornet/overview.html) page) is also a helpful resource.

Please kindly let us know what is confusing so that we can clarify it. Thanks 🙂

ps: Thanks for bringing it up. I noticed that not all samples that import `OptimizerOptions` use it. We'll get the redundancy pruned in a future release.

**Comment by NathanCQC** [comment]
Thanks for the very clear response. I was not aware of the interchangeability of data classes and dictionaries. Apologies for this.

I apologise for the frustration. Most of my frustration is from trying to do auto parrelism (I don't think the python contract docs mention this at all). I have had some problem running it on Perlmutter.

Hence I tried to write my own MPI slicing code, but this required careful control of all the slicing options and optimiser options etc., particularly for different processes. Even the ones that were not exposed in the python API. I have been stuck on this for over a week now and it is holding back my development. I have raised this in this discussion https://github.com/NVIDIA/cuQuantum/discussions/59. I am getting path length errors on different processes. Hence I was trying to freeze optimisation of internal tensors to see if that would help and this enum is not exposed in the python API.

Therefore it was hard to understand what a lot of the optimiser options do without examples. But I do agree they are well documented, however more examples would be clearer and help one understand quicker.

I want to be able to access the OptimiserOptions (and other) dataclasses even if they are wrapped up inside the Network.contract() method. Is this possible? The path and largest tensor data etc are very useful for analysis. 

Also you guys have done a great job with the code. Thanks a lot for all your hardwork.

**Reply by leofang** [reply]
> Most of my frustration is from trying to do auto parrelism (I don't think the python contract docs mention this at all). 

It's mentioned here: https://docs.nvidia.com/cuda/cuquantum/python/overview.html#id1 but indeed it is not always easy to navigate in docs. This is a common pain to all projects, not specific to us I'd think. Sometimes things are documented, but it's either hard for users to locate (from users perspective; for developers it's always like "isn't it already there?" 😅) or it's very easy to overlook. 

In any case, we're open for suggestions to make things better/easier for our users.

One useful tip I like with Sphinx-based documentation (like ours) is that there's a search bar that you can use. I often end up searching (if I know the keywords) instead of navigating.

> I have raised this in this discussion https://github.com/NVIDIA/cuQuantum/discussions/59. I am getting path length errors on different processes.

Yes it's addressed there. Thanks for keeping different topics separated in different discussion threads.

> I want to be able to access the OptimiserOptions (and other) dataclasses even if they are wrapped up inside the Network.contract() method. Is this possible? The path and largest tensor data etc are very useful for analysis.

`OptimizerOptions` is the user-provided option that you pass to functions/methods like `contract()` or `contract_path()`. If you need to access the information returned by the path optimizer, you can pass `return_info=True` to these APIs, and an `OptimizerInfo` object would be returned along with the computed results for you to inspect.
 
> Also you guys have done a great job with the code. Thanks a lot for all your hardwork.

Appreciated your feedbacks, Nathan! Do ask questions along the way, and we'll do our best to address them.

**Comment by leofang** [comment]
(Let me reopen this discussion. I noticed that if a discussion is marked as closed, at the entry time to GitHub Discussions it's not visible to readers, unless they specifically look for "closed" or "all" discussions.)
